mfsmaster/chunks.c

```diff
@@ -3521,28 +3521,33 @@ void chunk_do_jobs(chunk *c,uint16_t scount,uint16_t fullservers,uint32_t now,ui
 						}
 					}
 */
+
 					for (i=0 ; i<dstservcnt && canbefixed ; i++) {
 						if (matching[i+labelcnt]>=0 || allowallservers) {
-							uint32_t r;
-							if (rgvc>0) {	// if there are VALID copies then make copy of one VALID chunk
-								r = 1+rndu32_ranged(rgvc);
-								srccsid = MAXCSCOUNT;
-								for (s=c->slisthead ; s && r>0 ; s=s->next) {
-									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && s->valid==VALID) {
-										r--;
-										srccsid = s->csid;
-									}
-								}
-							} else {	// if not then use TDVALID chunks.
-								r = 1+rndu32_ranged(rgtdc);
-								srccsid = MAXCSCOUNT;
-								for (s=c->slisthead ; s && r>0 ; s=s->next) {
-									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && s->valid==TDVALID) {
-										r--;
-										srccsid = s->csid;
+							uint32_t min_dist = 0xFFFFFFFF;
+							uint32_t dist;
+							uint32_t ip;
+							uint32_t cuip;
+							uint16_t port;
+
+							srccsid = MAXCSCOUNT;
+
+							if (matocsserv_get_csdata(cstab[servers[i]].ptr,&cuip,&port,NULL,NULL)==0) {
+								for (s=c->slisthead ; s ; s=s->next) {
+									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass]) {
+										if (matocsserv_get_csdata(cstab[s->csid].ptr,&ip,&port,NULL,NULL)==0) {
+											dist=topology_distance(ip,cuip);
+											if (min_dist>dist || (min_dist==dist && s->valid==VALID)) {
```

Comment:
Hello @acid-maker 

> First in your code you do not use TDVALID chunks (sometimes in case of undergoal you may have only TDVALID chunks left - you should use them).

Actually I do use the TDVALID chunks, this `if` condition make it choose the minimum distance first, which may be a TDVALID chunk, if there's more than one minimum distance, choose the valid one.

>  Second in case of the same distance you should still randomize source.

I will fix the randomness, thanks.

Reply:
ok. you are right, but then you can use chunks in other states (i.e. INVALID copies) - you should add to one 'if' (s->valid==VALID || s->valid==TDVALID).

Reply:
You are right, I missed that.